elevate-cpanel

```diff
@@ -2888,6 +2905,238 @@ sub _latest_checksum () {
+        symlink $grub2_dir, $grub_dir or do {
+            rename $grub_bak, $grub_dir;                                                         # undo previous change on failure
+            LOGDIE("Unable to create symlink $grub_dir to point to $grub2_dir");                 # symlink() doesn't set $!
+        };
+
+    }
+
+    return;
+}
+
+sub merge_grub_directories_if_needed () {
+    my $stage_info = read_stage_file();
+    my $grub2_info = $stage_info->{'grub2_workaround'};
+    if ( $grub2_info && $grub2_info->{'needs_workaround_update'} ) {
+        my $grub_dir  = GRUB2_PREFIX_DEBIAN;
+        my $grub2_dir = GRUB2_PREFIX_RHEL;
+        my $grub_bak  = $grub2_info->{'backup_dir'};
+
+        my ( $skipped_copy, $failed_copy ) = ( 0, 0 );
+
+        # If something goes wrong here, we should give up on the copy but allow the upgrade to continue.
+        # Therefore, don't use LOGDIE() inside of the eval block.
+        eval {
+            my %grub2_entries;
+
+            opendir my $grub2_dir_fh, $grub2_dir or die "Unable to open directory $grub2_dir: $!";
+            while ( my $entry = readdir $grub2_dir_fh ) {
+                $grub2_entries{$entry} = 1;
+            }
+            closedir $grub2_dir_fh;
+
+            opendir my $grub_bak_fh, $grub_bak or die "Unable to open directory $grub_bak: $!";
+            while ( my $entry = readdir $grub_bak_fh ) {
+
+                next if $entry eq '.';
+                next if $entry eq '..';
+
+                # grub.cfg in the backup dir is the old symlink, so ignore it separately to avoid warnings:
+                next if $entry eq "grub.cfg";
+
+                if ( exists $grub2_entries{$entry} ) {
+                    $skipped_copy++;
+                    WARN("\"$grub_bak/$entry\" is not being copied to \"$grub2_dir/$entry\" because the destination already exists.");
+                    next;
+                }
+
+                if ( !File::Copy::Recursive::rcopy( "$grub_bak/$entry", "$grub2_dir/$entry" ) ) {
+                    $failed_copy++;
+                    WARN("Copying \"$grub_bak/$entry\" into \"$grub2_dir\" failed.");
+                }
+            }
+            closedir $grub_bak_fh;
+        };
+        WARN("Unable to copy the contents of \"$grub_bak\" into \"$grub2_dir\": $@") if $@;
+
+        my $log_file = LOG_FILE;
+        add_final_notification( <<~EOS ) if ( $skipped_copy > 0 || $failed_copy > 0 );
+        After converting "$grub_dir" from a directory to a symlink to
+        "$grub2_dir", the upgrade process chose not to copy $skipped_copy
+        entries from the old directory due to name conflicts, and it failed to
+        copy $failed_copy entries due to system errors. The previous contents
+        of "$grub_dir" are now located at "$grub_bak". See $log_file for
+        further information.
+
+        If you did not add these files or otherwise customize the boot loader
+        configuration, you may ignore this message.
```

Comment:
I'm not sure about "you may ignore this message." but I'm not sure what alternate language to suggest.

Reply:
My thinking here is that if not copying over a file breaks something non-critical with the bootloader, and the admin didn't take the action which caused the dependence on that file, there's no useful advice to give, thus it's better not to suggest anything to make it worse or otherwise confuse them. The final notification serves as notice to anyone who did perform weird customization that they should look to the log file to find out what files and issues occurred if something ancillary or cosmetic with the bootloader broke, rather than leaving them in the dark.